File: Server/models/InventoryItem.js
const mongoose = require('mongoose');

const inventoryItemSchema = new mongoose.Schema({
  // Basic Information
  name: {
    type: String,
    required: [true, 'Item name is required'],
    trim: true,
  },
  sku: {
    type: String,
    required: [true, 'SKU is required'],
    unique: true,
    uppercase: true,
    trim: true,
    index: true,
  },
  description: {
    type: String,
    trim: true,
  },
  
  // Categorization
  category: {
    type: String,
    required: [true, 'Category is required'],
    enum: ['Electronics', 'Clothing', 'Food', 'Furniture', 'Stationery', 'Medical', 'Automotive', 'Other'],
  },
  subCategory: {
    type: String,
    trim: true,
  },
  
  // Stock Information
  quantity: {
    type: Number,
    required: [true, 'Quantity is required'],
    min: [0, 'Quantity cannot be negative'],
    default: 0,
  },
  unit: {
    type: String,
    required: [true, 'Unit is required'],
    enum: ['pcs', 'kg', 'ltr', 'box', 'pack', 'dozen', 'meter'],
    default: 'pcs',
  },
  reorderLevel: {
    type: Number,
    required: [true, 'Reorder level is required'],
    min: [0, 'Reorder level cannot be negative'],
    default: 10,
  },
  maxStockLevel: {
    type: Number,
    min: [0, 'Max stock level cannot be negative'],
  },
  
  // Pricing
  costPrice: {
    type: Number,
    required: [true, 'Cost price is required'],
    min: [0, 'Cost price cannot be negative'],
  },
  sellingPrice: {
    type: Number,
    required: [true, 'Selling price is required'],
    min: [0, 'Selling price cannot be negative'],
  },
  
  // Supplier Information
  supplier: {
    name: {
      type: String,
      trim: true,
    },
    contactPerson: {
      type: String,
      trim: true,
    },
    phone: {
      type: String,
      trim: true,
    },
    email: {
      type: String,
      trim: true,
      lowercase: true,
    },
  },
  
  // Location & Storage
  warehouseLocation: {
    type: String,
    trim: true,
  },
  rackNumber: {
    type: String,
    trim: true,
  },
  
  // Expiry & Batch
  expiryDate: {
    type: Date,
  },
  batchNumber: {
    type: String,
    trim: true,
  },
  manufacturingDate: {
    type: Date,
  },
  
  // Status & Tracking
  status: {
    type: String,
    enum: ['active', 'inactive', 'discontinued'],
    default: 'active',
  },
  
  // Computed fields
  stockValue: {
    type: Number,
    default: 0,
  },
  
  // Image
  imageUrl: {
    type: String,
    trim: true,
  },
  
  // Metadata
  lastRestocked: {
    type: Date,
  },
  notes: {
    type: String,
    trim: true,
  },
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true },
});

// Virtual for stock status
inventoryItemSchema.virtual('stockStatus').get(function() {
  if (this.quantity === 0) return 'out-of-stock';
  if (this.quantity <= this.reorderLevel) return 'low-stock';
  if (this.maxStockLevel && this.quantity >= this.maxStockLevel) return 'overstock';
  return 'in-stock';
});

// Virtual for expiry status
inventoryItemSchema.virtual('expiryStatus').get(function() {
  if (!this.expiryDate) return null;
  
  const today = new Date();
  const daysUntilExpiry = Math.ceil((this.expiryDate - today) / (1000 * 60 * 60 * 24));
  
  if (daysUntilExpiry < 0) return 'expired';
  if (daysUntilExpiry <= 7) return 'expiring-soon';
  if (daysUntilExpiry <= 30) return 'expiring-this-month';
  return 'valid';
});

// Middleware to calculate stock value before saving
inventoryItemSchema.pre('save', function(next) {
  this.stockValue = this.quantity * this.costPrice;
  next();
});

// Index for better query performance
inventoryItemSchema.index({ category: 1, status: 1 });
inventoryItemSchema.index({ name: 'text', description: 'text' });

const InventoryItem = mongoose.model('InventoryItem', inventoryItemSchema);

module.exports = InventoryItem;

File: Server/models/inventory_item.py
from datetime import datetime
from bson.objectid import ObjectId
from config.database import get_db


class InventoryItem:
    """Inventory Item Model"""

    CATEGORIES = [
        'Electronics', 'Furniture', 'Clothing', 'Food & Beverage',
        'Raw Materials', 'Finished Goods', 'Office Supplies',
        'Medical', 'Automotive', 'Other'
    ]

    UNITS = ['pcs', 'kg', 'ltr', 'box', 'pack',
             'dozen', 'meter', 'carton', 'bag', 'roll']

    STATUSES = ['active', 'inactive', 'discontinued']

    @staticmethod
    def get_collection():
        """Get inventory items collection"""
        db = get_db()
        return db.inventoryitems  # type: ignore

    @staticmethod
    def create(data):
        """Create a new inventory item"""
        collection = InventoryItem.get_collection()

        # Calculate stock value
        data['stockValue'] = data.get('quantity', 0) * data.get('costPrice', 0)
        data['createdAt'] = datetime.utcnow()
        data['updatedAt'] = datetime.utcnow()

        # Set defaults
        if 'status' not in data:
            data['status'] = 'active'
        if 'unit' not in data:
            data['unit'] = 'pcs'
        if 'lastRestocked' not in data and data.get('quantity', 0) > 0:
            data['lastRestocked'] = datetime.utcnow()

        result = collection.insert_one(data)
        data['_id'] = result.inserted_id
        return data

    @staticmethod
    def find_all(filter_dict=None, skip=0, limit=10, sort_by='createdAt', sort_order=-1):
        """Find all inventory items with pagination"""
        collection = InventoryItem.get_collection()
        filter_dict = filter_dict or {}

        cursor = collection.find(filter_dict).sort(
            sort_by, sort_order).skip(skip).limit(limit)
        items = list(cursor)
        total = collection.count_documents(filter_dict)

        return items, total

    @staticmethod
    def find_by_id(item_id):
        """Find inventory item by ID"""
        collection = InventoryItem.get_collection()
        return collection.find_one({'_id': ObjectId(item_id)})

    @staticmethod
    def find_by_sku(sku):
        """Find inventory item by SKU"""
        collection = InventoryItem.get_collection()
        return collection.find_one({'sku': sku.upper()})

    @staticmethod
    def update(item_id, data):
        """Update inventory item"""
        collection = InventoryItem.get_collection()

        # Recalculate stock value if quantity or cost price changed
        if 'quantity' in data or 'costPrice' in data:
            item = InventoryItem.find_by_id(item_id)
            if item:
                quantity = data.get('quantity', item.get('quantity', 0))
                cost_price = data.get('costPrice', item.get('costPrice', 0))
                data['stockValue'] = quantity * cost_price

        data['updatedAt'] = datetime.utcnow()

        result = collection.update_one(
            {'_id': ObjectId(item_id)},
            {'$set': data}
        )

        if result.modified_count > 0:
            return InventoryItem.find_by_id(item_id)
        return None

    @staticmethod
    def delete(item_id):
        """Delete inventory item"""
        collection = InventoryItem.get_collection()
        result = collection.delete_one({'_id': ObjectId(item_id)})
        return result.deleted_count > 0

    @staticmethod
    def get_stock_status(item):
        """Calculate stock status for an item"""
        quantity = item.get('quantity', 0)
        reorder_level = item.get('reorderLevel', 0)
        max_stock_level = item.get('maxStockLevel')

        if quantity == 0:
            return 'out-of-stock'
        elif quantity <= reorder_level:
            return 'low-stock'
        elif max_stock_level and quantity >= max_stock_level:
            return 'overstock'
        else:
            return 'in-stock'

    @staticmethod
    def get_expiry_status(item):
        """Calculate expiry status for an item"""
        expiry_date = item.get('expiryDate')
        if not expiry_date:
            return None

        today = datetime.utcnow()
        if isinstance(expiry_date, str):
            expiry_date = datetime.fromisoformat(
                expiry_date.replace('Z', '+00:00'))

        days_until_expiry = (expiry_date - today).days

        if days_until_expiry < 0:
            return 'expired'
        elif days_until_expiry <= 7:
            return 'expiring-soon'
        elif days_until_expiry <= 30:
            return 'expiring-this-month'
        else:
            return 'valid'

    @staticmethod
    def to_dict(item):
        """Convert MongoDB document to dictionary with virtuals"""
        if not item:
            return None

        # Convert ObjectId to string
        item['id'] = str(item['_id'])
        del item['_id']  # Remove the original ObjectId field

        # Convert dates to ISO format strings
        if 'createdAt' in item and isinstance(item['createdAt'], datetime):
            item['createdAt'] = item['createdAt'].isoformat()
        if 'updatedAt' in item and isinstance(item['updatedAt'], datetime):
            item['updatedAt'] = item['updatedAt'].isoformat()
        if 'expiryDate' in item and isinstance(item['expiryDate'], datetime):
            item['expiryDate'] = item['expiryDate'].isoformat()
        if 'manufacturingDate' in item and isinstance(item['manufacturingDate'], datetime):
            item['manufacturingDate'] = item['manufacturingDate'].isoformat()
        if 'lastRestocked' in item and isinstance(item['lastRestocked'], datetime):
            item['lastRestocked'] = item['lastRestocked'].isoformat()

        item['stockStatus'] = InventoryItem.get_stock_status(item)
        item['expiryStatus'] = InventoryItem.get_expiry_status(item)

        return item

    @staticmethod
    def create_indexes():
        """Create database indexes"""
        collection = InventoryItem.get_collection()

        # Create indexes for better performance
        collection.create_index('sku', unique=True)
        collection.create_index([('category', 1), ('status', 1)])
        collection.create_index([('name', 'text'), ('description', 'text')])

        print('✅ Inventory item indexes created')
